File: arrayAndHashing9/validAnagram.java
package arrayAndHashing9;

import java.util.Arrays;


//Approach 3
//Time complexity: O(n log n)
//Space complexity: O(n)
class validAnagram {
   
    public boolean isAnagram(String s, String t) {
       String[] c1 = s.split("");
      String[] c2 = t.split("");
      Arrays.sort(c1);
      Arrays.sort(c2);
      return Arrays.equals(c1, c2);
    }
    public static void main(String[] args) {
        validAnagram obj = new validAnagram();
        String s = "anagram";
        String t = "nagaram";
        System.out.println(obj.isAnagram(s, t));
    }
}











// aproach 2
//Time complexity: O(n log n)
//Space complexity: O(n)
/*
class validAnagram {
   
    public boolean isAnagram(String s, String t) {
       if(s.length()!= t.length()){
           return false;
       }
       String [] array1 = s.split("");
       Arrays.sort(array1);
       String [] array2 = t.split("");
       Arrays.sort(array2);
       for(int i=0; i<array1.length;i++){
           if(!array1[i].equals(array2[i])){
               return false;
           }
       }
       return true;
       
    }
    public static void main(String[] args) {
        validAnagram obj = new validAnagram();
        String s = "anagram";
        String t = "nagaram";
        System.out.println(obj.isAnagram(s, t));
    }
}

*/




//Approach 1
/*
//Time complexity: O(n log n)
//Space complexity: O(n)
class validAnagram {

    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;

        }

        char[] c1 = s.toLowerCase().toCharArray();
        char[] c2 = t.toLowerCase().toCharArray();
        Arrays.sort(c1);
        Arrays.sort(c2);
        for (int i = 0; i < c1.length; i++) {
            if (c1[i] != c2[i]) {
                return false;

            }
        }
        return true;

    }

    public static void main(String[] args) {
        validAnagram obj = new validAnagram();
        String s = "anagram";
        String t = "nagaram";
        System.out.println(obj.isAnagram(s, t));
    }
}
*/

File: LEETCODE128/Valid Anagram.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LEETCODE128
{



    // Approach 2 
    // TC O(n)
    // SC O(1)
    internal class Valid_Anagram
    {
        public bool IsAnagram(string s, string t)
        {
            int a = s.Length;
            int b = t.Length;
            if (a != b)
            {
                return false;
            }
            int[] c = new int[26];
            int[] d = new int[26];
            for (int i = 0; i < a; i++)
            {
                c[s[i] - 'a']++;
            }
            for (int i = 0; i < a; i++)
            {
                d[t[i] - 'a']++;
            }
            if (c.SequenceEqual(d))
            {
                return true;
            }
            return false;
        }

        /*
        static void Main(string[] args)
        {
            Valid_Anagram obj = new Valid_Anagram();

            string s = "anagram";
            string t = "nagaram";
            Console.WriteLine(obj.IsAnagram(s, t));


        }

        */
    }

    /*

    //TC O(nlogn)
   // OC O(n)
    internal class Valid_Anagram
    {
        public bool IsAnagram(string s, string t)
        {
            if(s.Length != t.Length) return false;

            char[] c1  = s.ToCharArray();
            char[] c2 = t.ToCharArray();    
             Array.Sort(c1);
              Array.Sort(c2);
            for(int i = 0; i < c1.Length; i++)
            {
                if (c1[i] != c2[i]) return false;
               
                    
                
            }
            return true;

        }

        static void Main(string[] args)
        {
            Valid_Anagram obj = new Valid_Anagram();

            string s = "anagram";
            string t = "nagaram";
            Console.WriteLine(obj.IsAnagram(s,t));


        }
    }

    */
}
